File: apps/mobile/android/wear/src/main/java/app/marque/wear/ScoreEngine.kt
package app.marque.wear

import kotlinx.serialization.Serializable
import kotlin.math.ceil
import kotlin.math.max
import kotlin.math.min

/**
 * Full, config-driven port of the canonical TypeScript scoring engine. Rules are
 * DATA (RuleSetConfig) exactly as on the phone, so Wear OS handles every deuce
 * mode, tiebreaks with correct serve rotation, super-tiebreak deciders, mini-sets
 * and best-of — with zero divergence. Validated against the shared golden-vector
 * fixture in ScoreEngineTest (the enforceable phone/watch parity contract).
 */

@Serializable
data class PointConfig(
    val ladder: List<String>,
    val winAtIndex: Int,
    val pointMargin: Int,
    val deuce: String, // advantage | golden | star
    val starMaxAdvantages: Int? = null,
)

@Serializable
data class TiebreakConfig(
    val targetPoints: Int,
    val pointMargin: Int,
    val firstServerPoints: Int,
    val serveEvery: Int,
    val changeEndsEvery: Int,
)

@Serializable
data class SetConfig(
    val gamesToWin: Int,
    val gameMargin: Int,
    val tiebreakAtGames: Int? = null,
    val tiebreak: TiebreakConfig? = null,
    val maxGames: Int? = null,
)

@Serializable
data class FinalSetConfig(val kind: String, val set: SetConfig? = null, val superTiebreak: TiebreakConfig? = null)

@Serializable
data class MatchConfig(
    val bestOf: Int,
    val regularSet: SetConfig,
    val finalSet: FinalSetConfig,
    val playOutAfterMatchPoint: Boolean,
)

@Serializable
data class ServeConfig(val format: String, val gameServeCycle: List<Int>, val slotSide: List<Int>)

@Serializable
data class RuleSetConfig(
    val id: String,
    val version: Int,
    val point: PointConfig,
    val match: MatchConfig,
    val serve: ServeConfig,
)

@Serializable
data class TiebreakScore(val points: List<Int>, val winner: Int? = null)

@Serializable
data class SetScore(
    val games: List<Int>,
    val tiebreak: TiebreakScore? = null,
    val isTiebreak: Boolean = false,
    val winner: Int? = null,
    val kind: String, // regular | final
)

@Serializable
sealed class Outcome {
    @Serializable object InProgress : Outcome()
    @Serializable data class Completed(val winner: Int) : Outcome()
    @Serializable data class Retired(val winner: Int, val retiree: Int) : Outcome()

    val winnerOrNull: Int?
        get() = when (this) {
            is Completed -> winner
            is Retired -> winner
            else -> null
        }
}

@Serializable
data class MatchState(
    val players: List<String>,
    val sets: List<SetScore>,
    val setsWon: List<Int>,
    val currentSetIndex: Int,
    val currentGame: List<Int>,
    val gamesStartedTotal: Int,
    val serverSlot: Int,
    val servingSide: Int,
    val outcome: Outcome,
    val complete: Boolean,
    val seq: Int,
)

sealed class Action {
    data class PointTo(val side: Int) : Action()
    object Replay : Action()
    data class Penalty(val side: Int, val unit: String) : Action()
    data class Retire(val side: Int) : Action()
    object Undo : Action()
}

class ScoreEngine(private val cfg: RuleSetConfig) {
    private fun other(s: Int) = if (s == 0) 1 else 0
    private val setsNeeded get() = ceil(cfg.match.bestOf / 2.0).toInt()

    private fun deciderThreshold(): Int = when (cfg.point.deuce) {
        "advantage" -> Int.MAX_VALUE
        "golden" -> cfg.point.winAtIndex - 1
        else -> cfg.point.winAtIndex - 1 + (cfg.point.starMaxAdvantages ?: 0)
    }

    private fun resolveGamePoint(pts: List<Int>): Int? {
        val hi = max(pts[0], pts[1]); val lo = min(pts[0], pts[1])
        val lead = hi - lo; val leader = if (pts[0] > pts[1]) 0 else 1
        if (lo >= deciderThreshold()) return if (lead >= 1) leader else null
        if (hi >= cfg.point.winAtIndex && lead >= cfg.point.pointMargin) return leader
        return null
    }

    private fun tiebreakWinner(pts: List<Int>, tb: TiebreakConfig): Int? {
        val hi = max(pts[0], pts[1]); val lead = kotlin.math.abs(pts[0] - pts[1])
        return if (hi >= tb.targetPoints && lead >= tb.pointMargin) (if (pts[0] > pts[1]) 0 else 1) else null
    }

    private fun tiebreakServeCursor(opener: Int, played: Int, tb: TiebreakConfig, len: Int): Int {
        if (played < tb.firstServerPoints) return opener % len
        val swaps = (played - tb.firstServerPoints) / tb.serveEvery + 1
        return (opener + swaps) % len
    }

    private fun configForSet(kind: String): SetConfig {
        if (kind == "regular") return cfg.match.regularSet
        val fs = cfg.match.finalSet
        return when (fs.kind) {
            "full" -> fs.set ?: cfg.match.regularSet
            "miniSet" -> fs.set ?: SetConfig(4, 2, 4, cfg.match.regularSet.tiebreak)
            else -> SetConfig(1, 1, 0, fs.superTiebreak)
        }
    }

    private fun openSet(index: Int): SetScore {
        val kind = if (index == cfg.match.bestOf - 1) "final" else "regular"
        val setCfg = configForSet(kind)
        val startsTB = setCfg.tiebreakAtGames == 0
        return SetScore(
            games = listOf(0, 0),
            tiebreak = if (startsTB) TiebreakScore(listOf(0, 0)) else null,
            isTiebreak = startsTB, winner = null, kind = kind,
        )
    }

    private fun withServer(s: MatchState): MatchState {
        val len = cfg.serve.gameServeCycle.size
        val set = s.sets[s.currentSetIndex]
        val slot: Int
        if (set.isTiebreak && set.tiebreak != null) {
            val tbCfg = configForSet(set.kind).tiebreak!!
            val opener = s.gamesStartedTotal % len
            val played = set.tiebreak.points[0] + set.tiebreak.points[1]
            slot = cfg.serve.gameServeCycle[tiebreakServeCursor(opener, played, tbCfg, len) % len]
        } else {
            slot = cfg.serve.gameServeCycle[s.gamesStartedTotal % len]
        }
        return s.copy(serverSlot = slot, servingSide = cfg.serve.slotSide[slot])
    }

    fun initialState(players: List<String>): MatchState = withServer(
        MatchState(
            players = players, sets = listOf(openSet(0)), setsWon = listOf(0, 0),
            currentSetIndex = 0, currentGame = listOf(0, 0), gamesStartedTotal = 0,
            serverSlot = cfg.serve.gameServeCycle[0], servingSide = 0,
            outcome = Outcome.InProgress, complete = false, seq = 0,
        )
    )

    private fun replaceSet(s: MatchState, set: SetScore): MatchState {
        val sets = s.sets.toMutableList(); sets[s.currentSetIndex] = set
        return s.copy(sets = sets)
    }

    private fun closeSet(s0: MatchState, winner: Int, gamesStartedTotal: Int): MatchState {
        val setsWon = s0.setsWon.toMutableList(); setsWon[winner] += 1
        var s = s0.copy(setsWon = setsWon, gamesStartedTotal = gamesStartedTotal, currentGame = listOf(0, 0))
        val matchWon = setsWon[winner] >= setsNeeded
        if (matchWon) {
            s = s.copy(outcome = Outcome.Completed(winner), complete = true)
            if (!cfg.match.playOutAfterMatchPoint) return s
        }
        val nextIndex = s0.currentSetIndex + 1
        return withServer(s.copy(sets = s.sets + openSet(nextIndex), currentSetIndex = nextIndex))
    }

    private fun closeGame(s: MatchState, winner: Int): MatchState {
        val set = s.sets[s.currentSetIndex]
        val setCfg = configForSet(set.kind)
        val games = set.games.toMutableList(); games[winner] += 1
        val gamesStartedTotal = s.gamesStartedTotal + 1

        val tbAt = setCfg.tiebreakAtGames
        if (tbAt != null && tbAt > 0 && games[0] == tbAt && games[1] == tbAt) {
            val ns = set.copy(games = games, isTiebreak = true, tiebreak = TiebreakScore(listOf(0, 0)))
            return withServer(replaceSet(s.copy(gamesStartedTotal = gamesStartedTotal, currentGame = listOf(0, 0)), ns))
        }
        val lead = games[winner] - games[other(winner)]
        var won = games[winner] >= setCfg.gamesToWin && lead >= setCfg.gameMargin
        if (setCfg.maxGames != null && games[winner] >= setCfg.maxGames) won = true
        if (!won) {
            val ns = set.copy(games = games)
            return withServer(replaceSet(s.copy(gamesStartedTotal = gamesStartedTotal, currentGame = listOf(0, 0)), ns))
        }
        val ns = set.copy(games = games, winner = winner)
        return closeSet(replaceSet(s, ns), winner, gamesStartedTotal)
    }

    private fun tiebreakPoint(s: MatchState, side: Int): MatchState {
        val set = s.sets[s.currentSetIndex]
        val tbCfg = configForSet(set.kind).tiebreak!!
        val pts = set.tiebreak!!.points.toMutableList(); pts[side] += 1
        val winner = tiebreakWinner(pts, tbCfg)
        if (winner != null) {
            val games = set.games.toMutableList(); games[winner] += 1
            val gamesStartedTotal = s.gamesStartedTotal + 1
            val ns = set.copy(games = games, tiebreak = TiebreakScore(pts, winner), isTiebreak = false, winner = winner)
            return closeSet(replaceSet(s, ns), winner, gamesStartedTotal)
        }
        return withServer(replaceSet(s, set.copy(tiebreak = TiebreakScore(pts, null))))
    }

    private fun awardPoint(s: MatchState, side: Int): MatchState {
        val set = s.sets[s.currentSetIndex]
        if (set.isTiebreak) return tiebreakPoint(s, side)
        val pts = s.currentGame.toMutableList(); pts[side] += 1
        return if (resolveGamePoint(pts) != null) closeGame(s, side) else s.copy(currentGame = pts)
    }

    private fun isFrozen(s: MatchState): Boolean {
        if (!s.complete) return false
        if (cfg.match.playOutAfterMatchPoint && s.outcome is Outcome.Completed) return false
        return true
    }

    fun reduce(s: MatchState, action: Action): MatchState {
        fun bump(x: MatchState) = x.copy(seq = s.seq + 1)
        return when (action) {
            is Action.Undo -> s
            is Action.Replay -> bump(s)
            is Action.Retire ->
                if (s.outcome is Outcome.InProgress)
                    bump(s.copy(outcome = Outcome.Retired(other(action.side), action.side), complete = true))
                else bump(s)
            is Action.Penalty -> {
                if (isFrozen(s)) return bump(s)
                val beneficiary = other(action.side)
                val set = s.sets[s.currentSetIndex]
                if (action.unit == "game" && !set.isTiebreak) bump(closeGame(s, beneficiary))
                else bump(awardPoint(s, beneficiary))
            }
            is Action.PointTo -> if (isFrozen(s)) bump(s) else bump(awardPoint(s, action.side))
        }
    }
}

File: apps/mobile/ios/WatchApp/ScoreEngine.swift
import Foundation

/// Full, config-driven port of the canonical TypeScript scoring engine
/// (`packages/scoring-engine`). Rules are DATA (RuleSetConfig) exactly as on the
/// phone, so the watch handles every deuce mode, tiebreaks with correct serve
/// rotation, super-tiebreak deciders, mini-sets and best-of — with zero divergence.
///
/// This port is validated against the shared golden-vector fixture
/// (`fixtures/golden-vectors.json`) in ScoreEngineTests — the enforceable
/// phone/watch parity contract. Keep it in lockstep with the TS engine; when the
/// engine changes, regenerate the fixture and this test proves agreement.

// MARK: - Config (decodes the same JSON the phone stores)

enum DeuceMode: String, Codable { case advantage, golden, star }

struct PointConfig: Codable {
    let ladder: [String]
    let winAtIndex: Int
    let pointMargin: Int
    let deuce: DeuceMode
    let starMaxAdvantages: Int?
}

struct TiebreakConfig: Codable {
    let targetPoints: Int
    let pointMargin: Int
    let firstServerPoints: Int
    let serveEvery: Int
    let changeEndsEvery: Int
}

struct SetConfig: Codable {
    let gamesToWin: Int
    let gameMargin: Int
    let tiebreakAtGames: Int?
    let tiebreak: TiebreakConfig?
    let maxGames: Int?
}

struct FinalSetConfig: Codable {
    let kind: String // full | superTiebreak | miniSet
    let set: SetConfig?
    let superTiebreak: TiebreakConfig?
}

struct MatchConfig: Codable {
    let bestOf: Int
    let regularSet: SetConfig
    let finalSet: FinalSetConfig
    let playOutAfterMatchPoint: Bool
}

struct ServeConfig: Codable {
    let format: String
    let gameServeCycle: [Int]
    let slotSide: [Int]
}

struct RuleSetConfig: Codable {
    let id: String
    let version: Int
    let point: PointConfig
    let match: MatchConfig
    let serve: ServeConfig
}

// MARK: - State

struct TiebreakScore: Codable, Equatable { var points: [Int]; var winner: Int? }
struct GameScore: Codable, Equatable { var points: [Int]; var winner: Int? }

struct SetScore: Codable, Equatable {
    var games: [Int]
    var tiebreak: TiebreakScore?
    var isTiebreak: Bool
    var winner: Int?
    var kind: String // regular | final
}

enum Outcome: Equatable {
    case inProgress
    case completed(winner: Int)
    case retired(winner: Int, retiree: Int)
    var winner: Int? {
        switch self {
        case .inProgress: return nil
        case let .completed(w): return w
        case let .retired(w, _): return w
        }
    }
}

struct MatchState: Equatable {
    var players: [String]
    var sets: [SetScore]
    var setsWon: [Int]
    var currentSetIndex: Int
    var currentGame: GameScore
    var gamesStartedTotal: Int
    var serverSlot: Int
    var servingSide: Int
    var outcome: Outcome
    var complete: Bool
    var seq: Int
}

enum Action {
    case pointTo(Int)
    case replay
    case penalty(side: Int, unit: String)
    case retire(side: Int)
    case undo
}

// MARK: - Engine

struct ScoreEngine {
    let cfg: RuleSetConfig

    private func other(_ s: Int) -> Int { s == 0 ? 1 : 0 }
    private var setsNeeded: Int { Int(ceil(Double(cfg.match.bestOf) / 2.0)) }

    // points.ts
    private func deciderThreshold() -> Int {
        switch cfg.point.deuce {
        case .advantage: return Int.max
        case .golden: return cfg.point.winAtIndex - 1
        case .star: return cfg.point.winAtIndex - 1 + (cfg.point.starMaxAdvantages ?? 0)
        }
    }

    private func resolveGamePoint(_ pts: [Int]) -> Int? {
        let hi = max(pts[0], pts[1]); let lo = min(pts[0], pts[1])
        let lead = hi - lo; let leader = pts[0] > pts[1] ? 0 : 1
        if lo >= deciderThreshold() { return lead >= 1 ? leader : nil }
        if hi >= cfg.point.winAtIndex && lead >= cfg.point.pointMargin { return leader }
        return nil
    }

    // tiebreak.ts
    private func tiebreakWinner(_ pts: [Int], _ tb: TiebreakConfig) -> Int? {
        let hi = max(pts[0], pts[1]); let lead = abs(pts[0] - pts[1])
        if hi >= tb.targetPoints && lead >= tb.pointMargin { return pts[0] > pts[1] ? 0 : 1 }
        return nil
    }

    private func tiebreakServeCursor(_ opener: Int, _ played: Int, _ tb: TiebreakConfig, _ len: Int) -> Int {
        if played < tb.firstServerPoints { return opener % len }
        let swaps = (played - tb.firstServerPoints) / tb.serveEvery + 1
        return (opener + swaps) % len
    }

    // engine.ts: configForSet
    private func configForSet(_ kind: String) -> SetConfig {
        if kind == "regular" { return cfg.match.regularSet }
        let fs = cfg.match.finalSet
        switch fs.kind {
        case "full": return fs.set ?? cfg.match.regularSet
        case "miniSet":
            return fs.set ?? SetConfig(gamesToWin: 4, gameMargin: 2, tiebreakAtGames: 4,
                                       tiebreak: cfg.match.regularSet.tiebreak, maxGames: nil)
        default: // superTiebreak
            return SetConfig(gamesToWin: 1, gameMargin: 1, tiebreakAtGames: 0,
                             tiebreak: fs.superTiebreak, maxGames: nil)
        }
    }

    private func openSet(_ index: Int) -> SetScore {
        let kind = index == cfg.match.bestOf - 1 ? "final" : "regular"
        let setCfg = configForSet(kind)
        let startsTB = setCfg.tiebreakAtGames == 0
        return SetScore(games: [0, 0],
                        tiebreak: startsTB ? TiebreakScore(points: [0, 0], winner: nil) : nil,
                        isTiebreak: startsTB, winner: nil, kind: kind)
    }

    private func withServer(_ s: MatchState) -> MatchState {
        var st = s
        let len = cfg.serve.gameServeCycle.count
        let set = st.sets[st.currentSetIndex]
        if set.isTiebreak, let tb = set.tiebreak {
            let tbCfg = configForSet(set.kind).tiebreak!
            let opener = st.gamesStartedTotal % len
            let played = tb.points[0] + tb.points[1]
            let cursor = tiebreakServeCursor(opener, played, tbCfg, len)
            let slot = cfg.serve.gameServeCycle[cursor % len]
            st.serverSlot = slot; st.servingSide = cfg.serve.slotSide[slot]
        } else {
            let slot = cfg.serve.gameServeCycle[st.gamesStartedTotal % len]
            st.serverSlot = slot; st.servingSide = cfg.serve.slotSide[slot]
        }
        return st
    }

    func initialState(_ players: [String]) -> MatchState {
        var base = MatchState(players: players, sets: [openSet(0)], setsWon: [0, 0],
                              currentSetIndex: 0, currentGame: GameScore(points: [0, 0], winner: nil),
                              gamesStartedTotal: 0, serverSlot: cfg.serve.gameServeCycle[0], servingSide: 0,
                              outcome: .inProgress, complete: false, seq: 0)
        base = withServer(base)
        return base
    }

    private func closeSet(_ s: MatchState, _ winner: Int, _ gamesStartedTotal: Int) -> MatchState {
        var st = s
        st.setsWon[winner] += 1
        st.gamesStartedTotal = gamesStartedTotal
        st.currentGame = GameScore(points: [0, 0], winner: nil)
        let matchWon = st.setsWon[winner] >= setsNeeded
        if matchWon {
            st.outcome = .completed(winner: winner); st.complete = true
            if !cfg.match.playOutAfterMatchPoint { return st }
        }
        let nextIndex = s.currentSetIndex + 1
        st.sets.append(openSet(nextIndex))
        st.currentSetIndex = nextIndex
        return withServer(st)
    }

    private func closeGame(_ s: MatchState, _ winner: Int) -> MatchState {
        var st = s
        let set = st.sets[st.currentSetIndex]
        let setCfg = configForSet(set.kind)
        var games = set.games; games[winner] += 1
        let gamesStartedTotal = st.gamesStartedTotal + 1

        if let tbAt = setCfg.tiebreakAtGames, tbAt > 0, games[0] == tbAt, games[1] == tbAt {
            st.gamesStartedTotal = gamesStartedTotal
            st.currentGame = GameScore(points: [0, 0], winner: nil)
            var ns = set; ns.games = games; ns.isTiebreak = true
            ns.tiebreak = TiebreakScore(points: [0, 0], winner: nil)
            st.sets[st.currentSetIndex] = ns
            return withServer(st)
        }

        let lead = games[winner] - games[other(winner)]
        var won = games[winner] >= setCfg.gamesToWin && lead >= setCfg.gameMargin
        if let cap = setCfg.maxGames, games[winner] >= cap { won = true }

        if !won {
            st.gamesStartedTotal = gamesStartedTotal
            st.currentGame = GameScore(points: [0, 0], winner: nil)
            var ns = set; ns.games = games
            st.sets[st.currentSetIndex] = ns
            return withServer(st)
        }
        var ns = set; ns.games = games; ns.winner = winner
        st.sets[st.currentSetIndex] = ns
        return closeSet(st, winner, gamesStartedTotal)
    }

    private func tiebreakPoint(_ s: MatchState, _ side: Int) -> MatchState {
        var st = s
        let set = st.sets[st.currentSetIndex]
        let tbCfg = configForSet(set.kind).tiebreak!
        var pts = set.tiebreak!.points; pts[side] += 1
        if let winner = tiebreakWinner(pts, tbCfg) {
            var games = set.games; games[winner] += 1
            let gamesStartedTotal = st.gamesStartedTotal + 1
            var ns = set; ns.games = games; ns.tiebreak = TiebreakScore(points: pts, winner: winner)
            ns.isTiebreak = false; ns.winner = winner
            st.sets[st.currentSetIndex] = ns
            return closeSet(st, winner, gamesStartedTotal)
        }
        var ns = set; ns.tiebreak = TiebreakScore(points: pts, winner: nil)
        st.sets[st.currentSetIndex] = ns
        return withServer(st)
    }

    private func awardPoint(_ s: MatchState, _ side: Int) -> MatchState {
        let set = s.sets[s.currentSetIndex]
        if set.isTiebreak { return tiebreakPoint(s, side) }
        var pts = s.currentGame.points; pts[side] += 1
        if let winner = resolveGamePoint(pts) { _ = winner; return closeGame(s, side) }
        var st = s; st.currentGame = GameScore(points: pts, winner: nil); return st
    }

    private func isFrozen(_ s: MatchState) -> Bool {
        guard s.complete else { return false }
        if cfg.match.playOutAfterMatchPoint, case .completed = s.outcome { return false }
        return true
    }

    func reduce(_ s: MatchState, _ action: Action) -> MatchState {
        func bump(_ x: MatchState) -> MatchState { var y = x; y.seq = s.seq + 1; return y }
        switch action {
        case .undo: return s
        case .replay: return bump(s)
        case let .retire(side):
            if case .inProgress = s.outcome {
                var st = s; st.outcome = .retired(winner: other(side), retiree: side); st.complete = true
                return bump(st)
            }
            return bump(s)
        case let .penalty(side, unit):
            if isFrozen(s) { return bump(s) }
            let beneficiary = other(side)
            let set = s.sets[s.currentSetIndex]
            if unit == "game" && !set.isTiebreak { return bump(closeGame(s, beneficiary)) }
            return bump(awardPoint(s, beneficiary))
        case let .pointTo(side):
            if isFrozen(s) { return bump(s) }
            return bump(awardPoint(s, side))
        }
    }
}
